### PR TITLE
RFC: Create a module for picolibc

### LINKED ISSUE #44143: Adding picolibc as a module
## Origin

[Picolibc is available on github](https://github.com/picolibc/picolibc). The "official" upstream
repository is hosted in my own domain [keithp.com/picolibc](https://keithp.com/picolibc)

## Purpose

Picolibc provides a full libc implementation based on a combination of code from newlib, avr libc along with substantial new code, including the [ryu floating point conversion code](https://github.com/ulfjack/ryu) which provides malloc-free exact decimal/binary conversion of 32 and 64 bit IEEE floating point values.

## Mode of integration

As picolibc is used in numerous other projects, it should be integrated into zephyr as a module. 

## Pull Request

The PR is #44096 has been modified to take advantage of cmake support which has been added to picolibc.

## Description

Right now, Zephyr provides developers with several C library options, Newlib, Newlib-nano and a tiny 'minimal C library' included in tree. Picolibc fits in between these offerings with a full POSIX implementation like newlib/newlib-nano, but with a footprint similar to the included minimal library.

I think Picolibc could replace the current minimal C library, reducing the maintenance burden on the Zephyr project while offering users a better tested and more standards-conforming implementation, while potentially decreasing the amount of memory used by the resulting applications.

Picolibc could be made available through the Zephyr SDK as Crosstool-ng has the necessary support. However, that would mean selecting from the small number of binary configurations provided in that SDK, and would mean it wasn't available when using a different toolchain. Picolibc is routinely tested using GCC and Clang, and has been used with the CompCert verified C compiler. The codebase is tested using --std=c18 to ensure it remains in compliance with ISO/IEC 9899:2018.

Compared with newlib or newlib-nano, picolibc eliminates the 'struct _reent' by using TLS variables, replaces the large stdio stack with a lighter weight implementation which doesn't require malloc and has an integrated test suite which is used as part of the CI system to validate changes. Picolibc is also regularly tested using significant portions of the glibc test suite. All of that testing has led to dozens of bug fixes, many of which have been sent upstream to newlib, but many cannot as the picolibc code has diverged in significant ways, especially the math library.

Picolibc has very clear BSD licensing for the library source code; all of the non-BSD code included in the newlib repository has been removed. There are some tests and sample scripts which are not BSD licensed, but none of those are involved in building the library itself.

# Dependencies

The library itself only requires a C compiler (tested with GCC in C99, C11 and C18 modes), and Picolibc now has native cmake support. Picolibc is designed to work with systems that support thread local storage; all C library APIs which need per-thread storage use this mechanism, including `errno`.

## Revision

Picolibc includes a `zephyr` branch which has the necessary fixes and cmake support. When we're ready to merge this PR, that should shift to using a specific revision in the west configuration file.

## License

Picolibc is "BSD" licensed, but that includes numerous versions of BSD-2 and BSD-3 licenses, along with some other variants.

# Resources

* [picolibc footprint comparison](https://github.com/zephyrproject-rtos/zephyr/pull/26545#issue-648441961)
* [Rationale and overall requirements](https://github.com/picolibc/picolibc/issues/177#issuecomment-1073815655)
* [Configuration requirements](https://github.com/picolibc/picolibc/issues/177#issuecomment-1076039486)

### PR DESCRIPTION
**Pull request by keith-packard**
Rather than depending on the toolchain, create a module for picolibc and build it along with Zephyr. This has a few hacks to make it work; I'm mostly curious if this seems like a reasonable direction for picolibc integration. I've set it up to just use 'meson' to configure and build picolibc for now.

Closes #44143

### CODE REVIEW

**Comment by carlescufi** [comment]
Thanks @keith-packard, would you mind creating an issue as described [here](https://docs.zephyrproject.org/latest/contribute/external.html) so we can proceed with the creation of a mirror repo after a TSC vote?

**Comment by carlescufi** [comment]
Please ignore the "Manifest" job failure, that's a bug on the bot side.

**Comment by stephanosio** [comment]
Just to give some context to those who have not been involved in this conversation earlier:
https://github.com/picolibc/picolibc/issues/177#issuecomment-1073815655

**Review by stephanosio — COMMENTED** [review]
I have listed the configuration requirements for the picolibc to be able to replace the current Zephyr minimal libc in the following comment:
https://github.com/picolibc/picolibc/issues/177#issuecomment-1076039486

In general, we want more granular control over the internals of the picolibc as described in that comment.

Also, we would want to be able to choose (via Kconfig) the components of the picolibc to compile so that, when we build the Zephyr applications, we do not build the entire picolibc and unreasonably increase the build time. Note that this is important because, in case we decide to go through with replacing the Zephyr minimal libc with the picolibc, the picolibc will become the default libc for Zephyr.

**Comment by keith-packard** [comment]
I've mirrored the existing picolibc configuration options that are relevant to Zephyr into the Zephyr C Library configuration menu and apply those using the CMakeLists.txt provided on the zephyr branch of the picolibc repository. This ensures that the build mechanism used for picolibc is not exposed in the Zephyr repository, while still offering complete customization of the picolibc build.

**Comment by keith-packard** [comment]
I pushed some preliminary cmake bits to the [cmake branch of picolibc](https://github.com/picolibc/picolibc/tree/cmake). I'm a bit stuck though; I need to set some per-source compiler options and tried using this function:
```
function(picolibc_sources_flags flags)
  set(orig_sources ${ARGN})
  picolibc_sources(${orig_sources})
  foreach(source IN LISTS orig_sources)
    message(STATUS "set flags for ${CMAKE_CURRENT_SOURCE_DIR}/${source} to ${flags}")
    set_source_files_properties(${source}
      DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}"
      PROPERTIES COMPILE_FLAGS "${flags}")
  endforeach()
endfunction()
```
but I can't get it to work. I tried open-coding the call to set_source_files_properties and that also failed. I'm otherwise pretty close to building a working C library.

**Comment by keith-packard** [comment]
ok, I think I've got it building correctly, or at least can fix up any remaining issues. I'll see if I can't get zephyr to build tomorrow.

**Comment by keith-packard** [comment]
And picolibc now builds with cmake as a module under zephyr, with the result working on arm32, riscv32 and riscv64. All of the changes for that were done in the picolibc repository on the zephyr branch, which has been pushed to https://github.com/zephyrproject-rtos/picolibc

**Comment by keith-packard** [comment]
I've spent some quality time updating this patch series. At this point, it runs nearly all of the twister tests against the arm architecture.
```
$ ./scripts/twister --ninja -a arm >& twister.log
```
The memory protection tests continue to fail, and I can't figure them out -- they fault when attempting to access the z_arm_tls_ptr value, but the section containing that variable is included in the thread's list of MPU regions.

The picolibc changes to integrate cmake and zephyr support are moving upstream there; the zephyr branch contains all of the patches necessary to build zephyr against that. Advice on how to structure the zephyr-specific changes would be appreciated, it seems like most of the zephyr-specific files could be moved in to the zephyr tree?

**Comment by keith-packard** [comment]
And now I've cleaned up the Zephyr patch series so that it has all core Zephyr changes, followed by the Picolibc additions and finally changes to tests to make them work with Picolibc too. The last patch in this series changes the default C library from Minimal to Picolibc; I've included that to aid testing. With the patches here and the Zephyr branch from picolibc, which contains a few Picolibc fixes, cmake support and Zephyr configuration files, twister fails only on the memory protection tests which crash when the app tries to fetch the TLS pointer.

I've started working on RISC-V as well -- that also needs code to set the TP register when switching to user mode (there's a patch in the series above for 32-bit arm).

**Comment by keith-packard** [comment]
Ok, it looks like zephyr needs some TLS attention -- I've discovered numerous tests are failing when thread local storage is enabled, even when using the minimal C library. I've got patches for enough to make twister pass on -a arm and am checking -a riscv32 -a riscv64 to see how much more work is needed here.

This means that I believe picolibc support is now actually ready for final review -- the vast bulk of the patches in this series affect tests which need updating to support TLS. The actual picolibc changes are pretty minor. All of the necessary picolibc patches are moving upstream and are present on the zephyr branch.

**Comment by stephanosio** [comment]
Rebased

**Comment by stephanosio** [comment]
Will target 3.2.0 for this feature since the feature freeze for 3.1.0 is on this Friday.

**Review by stephanosio — COMMENTED** [review]
@keith-packard I think it would be helpful to split this series into two parts:

* Essential commits required for introducing the picolibc support to Zephyr (this PR)
* Extraneous commits that are not absolutely needed for supporting the picolibc in Zephyr (another PR); for example:
    * [lib/os: Use picolibc printf as a replacement for printk](https://github.com/zephyrproject-rtos/zephyr/pull/44096/commits/90478fc42fe8203ee1f3b63f88c0108edddd54e9)
    * [lib/os: Replace cbvprintf and printfcb family when using picolibc [v3]](https://github.com/zephyrproject-rtos/zephyr/pull/44096/commits/8e3473bc3366f75f3c595b1cdaf15f0ee606f2fb)

Some of the "extraneous commits" may be controversial and will likely take some time for everyone to reach a consensus; so, it would be good to get them reviewed in a separate PR so as to not block the introduction of the picolibc itself.

**Comment by keith-packard** [comment]
> @keith-packard I think it would be helpful to split this series into two parts:
> 
>     * Essential commits required for introducing the picolibc support to Zephyr (this PR)

Pushed to picolibc-module. I've also included the picolibc specific tests which don't depend on making picolibc the default C library, and the patches to those tests necessary to make them work.

>     * Extraneous commits that are not absolutely needed for supporting the picolibc in Zephyr (another PR); for example:
>       
>       * [lib/os: Use picolibc printf as a replacement for printk](https://github.com/zephyrproject-rtos/zephyr/pull/44096/commits/90478fc42fe8203ee1f3b63f88c0108edddd54e9)
>       * [lib/os: Replace cbvprintf and printfcb family when using picolibc [v3]](https://github.com/zephyrproject-rtos/zephyr/pull/44096/commits/8e3473bc3366f75f3c595b1cdaf15f0ee606f2fb)

I've pushed those to a 'picolibc-complete' branch; let's get picolibc merged and then I'll submit two separate PRs for those bits. 

> Some of the "extraneous commits" may be controversial and will likely take some time for everyone to reach a consensus; so, it would be good to get them reviewed in a separate PR so as to not block the introduction of the picolibc itself.

I hope not -- they both replicate existing APIs while making them smaller with better POSIX conformance. But, we'll see.

File: lib/libc/picolibc/CMakeLists.txt
Comment (stephanosio) [inline lib/libc/picolibc/CMakeLists.txt:6]: `if(CONFIG_PICOLIBC` is redundant because the `picolibc` subdirectory is only added when `CONFIG_PICOLIBC=y`.

File: lib/libc/picolibc/CMakeLists.txt
Comment (stephanosio) [inline lib/libc/picolibc/CMakeLists.txt:10]: For `-D_POSIX_C_SOURCE=200809`:
```
zephyr_compile_definitions(_POSIX_C_SOURCE=200809)
```

For `--specs=picolibc.specs`:
```
  zephyr_link_libraries(-specs=picolibc.specs)
  zephyr_compile_options(-specs=picolibc.specs)
```

File: lib/libc/picolibc/CMakeLists.txt
Comment (stephanosio) [inline lib/libc/picolibc/CMakeLists.txt:14]:
```
zephyr_compile_definitions(PICOLIBC_DOUBLE_PRINTF_SCANF)
```

File: lib/libc/picolibc/CMakeLists.txt
Comment (stephanosio) [inline lib/libc/picolibc/CMakeLists.txt:17]:
```
zephyr_compile_definitions(PICOLIBC_INTEGER_PRINTF_SCANF)
```

File: west.yml
Comment (stephanosio) [inline west.yml:204]: This needs to be pointing to a specific commit.

File: tests/lib/c_lib/src/main.c
Comment (stephanosio) [inline tests/lib/c_lib/src/main.c:1071]:
```
#else
	ztest_test_skip();
```

File: tests/lib/c_lib/src/main.c
Comment (stephanosio) [inline tests/lib/c_lib/src/main.c:1087]:
```
#else
	ztest_test_skip();
```

File: tests/lib/c_lib/src/main.c
Comment (stephanosio) [inline tests/lib/c_lib/src/main.c:1119]:
```
#else
	ztest_test_skip();
```

**Review by stephanosio — CHANGES_REQUESTED** [review]

**Comment by stephanosio** [comment]
@keith-packard How do we ensure thread-safety with the picolibc (e.g. calling `malloc` from multiple threads)?

At first glance, it seems like the picolibc functions will not be thread safe because the `libc-hooks.c` for it is missing the retargetable locking interface implementation, is that correct?

File: include/zephyr/sys/libc-hooks.h
Comment (keith-packard) [reply include/zephyr/sys/libc-hooks.h:70]: The code style checks disagree with you here; I suspect there's a bug in that.

File: include/zephyr/sys/libc-hooks.h
Comment (keith-packard) [reply include/zephyr/sys/libc-hooks.h:84]: same issue -- code style requires no indentation on these continuation lines.

File: west.yml
Comment (keith-packard) [reply west.yml:204]: Thanks; I was leaving it pointing to a branch so I didn't need to update zephyr when I made changes in picolibc. Now that there's a released version of picolibc with the required changes, I'll switch to a commit instead. I assume you don't want a tag name here?

Comment (stephanosio) [reply west.yml:204]: > I assume you don't want a tag name here?

While specifying a tag name should work, commit SHA would be preferred.

[We are in the process of cleaning up the upstream module repositories of unneeded branches (and tags)](https://github.com/zephyrproject-rtos/zephyr/issues/45241), so using commit SHA would be required as part of that too.

**Comment by keith-packard** [comment]
> @keith-packard How do we ensure thread-safety with the picolibc (e.g. calling `malloc` from multiple threads)?
> 
> At first glance, it seems like the picolibc functions will not be thread safe because the `libc-hooks.c` for it is missing the retargetable locking interface implementation, is that correct?

Yes. I borrowed the newlib hooks (amended for picolibc) in the most recent update.

File: lib/libc/picolibc/CMakeLists.txt
Comment (keith-packard) [reply lib/libc/picolibc/CMakeLists.txt:10]: Thanks, fixed.

File: lib/libc/picolibc/CMakeLists.txt
Comment (keith-packard) [reply lib/libc/picolibc/CMakeLists.txt:17]: Fixed

File: west.yml
Comment (keith-packard) [reply west.yml]: Fixed. Points to the 1.7.7 commit sha.

File: lib/libc/picolibc/CMakeLists.txt
Comment (keith-packard) [reply lib/libc/picolibc/CMakeLists.txt:6]: Fixed.

File: lib/libc/picolibc/CMakeLists.txt
Comment (keith-packard) [reply lib/libc/picolibc/CMakeLists.txt:14]: Fixed.

**Comment by keith-packard** [comment]
> Pushed to picolibc-module. I've also included the picolibc specific tests which don't depend on making picolibc the default C library, and the patches to those tests necessary to make them work.

I've also pushed a picolibc-fixes branch, which has all of the test/sample fixes needed to run the whole test suite using picolibc (on supported platforms).

Finally, I pushed picolibc-printk and picolibc-cbprintf which have the patches to enable those two features on top of the picolibc-module commit. They conflict as they both touch libc-hooks.c, but they are otherwise independently reviewable.

File: lib/libc/picolibc/CMakeLists.txt
Comment (stephanosio) [reply lib/libc/picolibc/CMakeLists.txt:6]: It looks like this change was not pushed.

File: lib/libc/picolibc/CMakeLists.txt
Comment (stephanosio) [reply lib/libc/picolibc/CMakeLists.txt:10]: It looks like this change was not pushed.

File: lib/libc/picolibc/CMakeLists.txt
Comment (stephanosio) [reply lib/libc/picolibc/CMakeLists.txt:14]: It looks like this change was not pushed.

File: lib/libc/picolibc/CMakeLists.txt
Comment (stephanosio) [reply lib/libc/picolibc/CMakeLists.txt:17]: It looks like this change was not pushed.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (stephanosio) [reply CMakeLists.txt:948]: Thanks for clarifying. Yes, it would be good to have that patch back in this series.

Comment (keith-packard) [reply CMakeLists.txt:948]: Sounds good. I've rebased and amended this series; will push (and resolve) when local tests are complete.

Comment (keith-packard) [reply CMakeLists.txt]: Local tests completed; found a couple more bugs (PRs posted), then pushed this branch again.

**Comment by andyross** [comment]
Seeing this go by and nodding every time.  Not to step too far into the project governance hole, but... what exactly are we waiting for here?  This is more libc engagement and development than we seen in... basically forever.  What's preventing us from just pulling the trigger and merging, with an eye to making picolibc default?

**Comment by stephanosio** [comment]
> what exactly are we waiting for here?

Not much really. All requested changes seem to have been addressed, the TSC has already approved this, so nothing from the governance side of things either.

It should be good to go once the CI is green (there is a failure on `hifive_unleashed` for some reason):
https://github.com/zephyrproject-rtos/zephyr/runs/6956910072?check_suite_focus=true#step:12:1436

I will revisit this PR for review some time this week.

File: lib/libc/picolibc/libc-hooks.c
Comment (stephanosio) [inline lib/libc/picolibc/libc-hooks.c:440]: Zephyr provides `CODE_UNREACHABLE` as part of the toolchain abstraction layer.

File: tests/lib/mem_alloc/testcase.yaml
Comment (stephanosio) [inline tests/lib/mem_alloc/testcase.yaml:53]: `libraries.libc.picolibc.mem_alloc`

File: lib/libc/picolibc/libc-hooks.c
Comment (keith-packard) [reply lib/libc/picolibc/libc-hooks.c]: and CODE_UNREACHABLE was already used a few lines above this. Fixed.

File: tests/lib/mem_alloc/testcase.yaml
Comment (keith-packard) [reply tests/lib/mem_alloc/testcase.yaml]: Fixed.

**Comment by keith-packard** [comment]
> It should be good to go once the CI is green (there is a failure on `hifive_unleashed` for some reason): https://github.com/zephyrproject-rtos/zephyr/runs/6956910072?check_suite_focus=true#step:12:1436

Not sure how I can help with this -- I don't have an unleashed board any more. This used to pass, and I haven't changed the code in the patch for quite a while. Someone with the hardware will likely need to bisect the merge point against the last successful run, and that doesn't sound like much fun.

> I will revisit this PR for review some time this week.

Thanks. I'd like to get this off my todo list if possible.

**Comment by nashif** [comment]
> > It should be good to go once the CI is green (there is a failure on `hifive_unleashed` for some reason): [zephyrproject-rtos/zephyr/runs/6956910072?check_suite_focus=true#step:12:1436](https://github.com/zephyrproject-rtos/zephyr/runs/6956910072?check_suite_focus=true#step:12:1436)
> 
> Not sure how I can help with this -- I don't have an unleashed board any more. This used to pass, and I haven't changed the code in the patch for quite a while. Someone with the hardware will likely need to bisect the merge point against the last successful run, and that doesn't sound like much fun.

this is happening in the renode simulator, so I do not think you need the board if you can deal with renode. Have seen issues with timing and renode in the past, could be just a false positive and something unique to renode. @mgielda @pgielda  @PiotrZierhoffer  can you please take a look?

**Comment by keith-packard** [comment]
> this is happening in the renode simulator, so I do not think you need the board if you can deal with renode. Have seen issues with timing and renode in the past, could be just a false positive and something unique to renode. @mgielda @pgielda @PiotrZierhoffer can you please take a look?

I added some debugging to the code so I could just let the CI run and see what happened. It turns out the hifive_unleashed is configured with a lot of memory (0xf000_0000 bytes). The test is requesting an allocation of 0x7fff_ffff bytes, which fits within this heap. As the test is calling calloc, it then calls memset on the result, which takes "a while". I assume the test is just timing out at that point. I've added a fix to the mem_alloc test to make sure allocations which "should" fail actually will.

File: tests/lib/mem_alloc/src/main.c
Comment (stephanosio) [inline tests/lib/mem_alloc/src/main.c:35]: `__PTRDIFF_MAX__` is not standard and toolchain-specific.

`PTRDIFF_MAX`, which is part of the C standard from C99 and on, should be used.

File: tests/subsys/logging/log_stack/src/main.c
Comment (stephanosio) [inline tests/subsys/logging/log_stack/src/main.c:138]: > The amount of stack used has changed for some reason (toolchain updates?),
use the amounts reported by the test when it failed.

This is likely from the main branch (same failure observed there too) -- we have been discussing whether this test should be disabled because the stack usage tends to change from time to time and it does not make sense to make the CI fail because of this (FYI @nordic-krch).

For now, this patch should do.

**Review by stephanosio — COMMENTED** [review]
Thanks for investigating the issue. The fix looks reasonable to me.

Just one minor change in the fix.

**Comment by PiotrZierhoffer** [comment]
@nashif @keith-packard I didn't follow the whole conversation, but it seems that the current code passes as expected because the previous version was supposed to fail. Is that correct?

Indeed, allocating this amount of memory in Renode can take a moment, but admittedly it's not a usual use case.

File: tests/lib/mem_alloc/src/main.c
Comment (keith-packard) [reply tests/lib/mem_alloc/src/main.c]: Thanks, I've updated the patch (without rebasing) to use PTRDIFF_MAX unconditionally after including stdint.h

File: arch/Kconfig
Comment (stephanosio) [inline arch/Kconfig:552]: IMHO, this would better handled via `depends on ARC || ARM || ARM64 || ...` from the `config PICOLIBC` because `ARCH_HAS_*` are for the availability of the architecture/ABI-specific kernel features, but no strong opinion in this particular instance since the picolibc will soon become an integral part of Zephyr.

**Review by stephanosio — DISMISSED** [review]
LGTM

Thanks for being patient and coping with the lengthy review process.

File: arch/Kconfig
Comment (keith-packard) [reply arch/Kconfig:552]: Hrm. What I want is a way to know when the picolibc-specific tests can be enabled; this replaced a pile of architecture tests in the testcase.yaml files. I guess I could stick a config PICOLIBC_SUPPORTS_ARCH in lib/libc/Kconfig? Name suggestions welcome, just want to make sure we have a single place which determines whether picolibc is supported or not, instead of littering them around the codebase.

Comment (stephanosio) [reply arch/Kconfig:552]: > What I want is a way to know when the picolibc-specific tests can be enabled

In that case, adding `PICOLIBC_SUPPORTED` or similar that has `default y` and `depends on ARC || ARM || ...` in `lib/libc/Kconfig` should do.

Also `config PICOLIBC` should depend on `PICOLIBC_SUPPORTED` since it would not be reasonable to allow users to select the picolibc for the archs that are not tested.

One question I have is: do we really need this? Is picolibc known to not work with any of the architectures currently supported in Zephyr? (i.e. do any of the tests fail on them?)

Comment (keith-packard) [reply arch/Kconfig]: Yeah, that looks better -- stick 'config ARCH_HAS_PICOLIBC' in lib/libc/Kconfig. Let me know if there's a better name for this.

Comment (stephanosio) [reply arch/Kconfig]: Since `config ARCH_HAS_PICOLIBC` is no longer in `arch/Kconfig`, `PICOLIBC_SUPPORTED` or `PICOLIBC_SUPPORTS_ARCH` would be better.

Though, this symbol may very well be generalised to indicate whether the picolibc with the current configuration is supported in general (i.e. not just for the different archs, but also for toolchains); in that sense, `PICOLIBC_SUPPORTED` could be a better choice than `PICOLIBC_SUPPORTS_ARCH`; but for now, either would be ok really.

Comment (stephanosio) [reply arch/Kconfig]: Also still wondering:

> One question I have is: do we really need this? Is picolibc known to not work with any of the architectures currently supported in Zephyr? (i.e. do any of the tests fail on them?)

Comment (keith-packard) [reply arch/Kconfig]: > Since `config ARCH_HAS_PICOLIBC` is no longer in `arch/Kconfig`, `PICOLIBC_SUPPORTED` or `PICOLIBC_SUPPORTS_ARCH` would be better.
> 
> Though, this symbol may very well be generalised to indicate whether the picolibc with the current configuration is supported in general (i.e. not just for the different archs, but also for toolchains); in that sense, `PICOLIBC_SUPPORTED` could be a better choice than `PICOLIBC_SUPPORTS_ARCH`; but for now, either would be ok really.

Sounds like `PICOLIBC_SUPPORTED` would be a better choice. I'll switch.

Comment (keith-packard) [reply arch/Kconfig]: > One question I have is: do we really need this? Is picolibc known to not work with any of the architectures currently supported in Zephyr? (i.e. do any of the tests fail on them?)

I tried x86_64 and there were some unexpected failures which I haven't debugged. Note that I'm doing much more significant testing off of the 'picolibc-fixes' branch, which switches the default C library to picolibc and then runs all of the tests, instead of only the subset included in the picolibc-module test. Small steps.

Comment (stephanosio) [reply arch/Kconfig]: > Note that I'm doing much more significant testing off of the 'picolibc-fixes' branch, which switches the default C library to picolibc and then runs all of the tests, instead of only the subset included in the picolibc-module test. Small steps.

Ok, thanks for clarifying. We can gradually address those issues after this PR is merged.

**Comment by keith-packard** [comment]
ok, I'll try to hold this version stable long enough to get a couple of approvals so we can just get 'er merged. If you have any comments about changes that are required, if they're relatively minor, let's plan on doing minor PRs after the merge to fix things up instead of delaying the merge.

**Review by andyross — APPROVED** [review]
Cool by me, certainly.  Nothing really impactful is happening in the Zephyr tree that I can see, mostly cleanup for existing weirdness.

**Review by stephanosio — APPROVED** [review]
LGTM

**Review by jhedberg — APPROVED** [review]

**Comment by enjiamai** [comment]
Hi @keith-packard,  the commit of the picolibc repo seems like a incorrect reversion in the west.yml.  It will cause west update failed. Could you please help on correcting it? thanks.

```
199     - name: picolibc
200       path: modules/lib/picolibc
201       revision: 4f8076520240192be284843a114c92853e29bbbc
```
I cannot find this commit:  4f8076520240192be284843a114c92853e29bbbc  in https://github.com/zephyrproject-rtos/picolibc

**Comment by enjiamai** [comment]
Update: it works after 0984aedf68e6ff967af93ee12ee7c5de81178ca6 got merged. thanks. Please ignore below comments, thanks!

**Comment by keith-packard** [comment]
> Update: it works after [0984aed](https://github.com/zephyrproject-rtos/zephyr/commit/0984aedf68e6ff967af93ee12ee7c5de81178ca6) got merged. thanks. Please ignore below comments, thanks!

Sorry for messing up the zephyr/picolibc repo -- looks like we've got a process in place now with the ability to FF the main branch.

**Pull request merged**
